packages/host/app/services/message-service.ts

```diff
@@ -1,35 +1,89 @@
+
 import Service, { service } from '@ember/service';
 
 import { tracked } from '@glimmer/tracking';
 
+import window from 'ember-window-mock';
+
+import qs from 'qs';
+
+import { sessionLocalStorageKey } from './realm';
+
+import type MatrixService from './matrix-service';
 import type NetworkService from './network';
+import type ResetService from './reset';
 
 export default class MessageService extends Service {
-  @tracked subscriptions: Map<string, EventSource> = new Map();
+  @tracked subscriptions: Map<
+    string,
+    { eventSource: EventSource; unsubscribes: (() => void)[] }
+  > = new Map();
   @service private declare network: NetworkService;
+  @service private declare reset: ResetService;
+  @service private declare matrixService: MatrixService;
+
+  constructor(owner: Owner) {
+    super(owner);
+    this.reset.register(this);
+  }
 
   register() {
     (globalThis as any)._CARDSTACK_REALM_SUBSCRIBE = this;
   }
 
-  subscribe(realmURL: string, cb: (ev: MessageEvent) => void): () => void {
-    let maybeEventSource = this.subscriptions.get(realmURL);
+  subscribe(realmURL: URL, cb: (ev: MessageEvent) => void): () => void {
+    let realm = realmURL.href;
+    if (!this.matrixService.isLoggedIn) {
+      throw new Error(
+        `Cannot subscribe to events from realm ${realm} before logging in `,
+      );
+    }
+    let { eventSource: maybeEventSource, unsubscribes = [] } =
+      this.subscriptions.get(realm) ?? {};
 
+    let eventSource: EventSource;
     if (!maybeEventSource) {
-      maybeEventSource = this.network.createEventSource(realmURL);
-      maybeEventSource.onerror = () => eventSource.close();
-      this.subscriptions.set(realmURL, maybeEventSource);
+      let token = getPersistedTokenForRealm(realm);
+      if (!token) {
+        throw new Error(`Could not find JWT for realm ${realm}`);
+      }
+      let urlWithAuth = `${realm}_message?${qs.stringify({
+        authHeader: 'Bearer ' + token,
+      })}`;
+      eventSource = this.network.createEventSource(urlWithAuth);
+      eventSource.onerror = () => eventSource!.close();
+      this.subscriptions.set(realm, { eventSource, unsubscribes });
+    } else {
+      eventSource = maybeEventSource;
     }
 
-    let eventSource = maybeEventSource;
     // TODO might want to consider making separate subscription methods so that
     // you can subscribe to a specific type of events instead of all of the
     // events...
     eventSource.addEventListener('update', cb);
     eventSource.addEventListener('index', cb);
-    return () => {
+
+    let unsubscribe = () => {
       eventSource.removeEventListener('update', cb);
       eventSource.removeEventListener('index', cb);
     };
+    unsubscribes.push(unsubscribe);
+    return unsubscribe;
+  }
+
+  resetState() {
+    for (let { unsubscribes } of this.subscriptions.values()) {
+      for (let unsubscribe of unsubscribes) {
+        unsubscribe();
+      }
+    }
+    this.subscriptions = new Map();
```

Comment:
i think it was an omission that we never had a way to unsubscribe from SSE when logging out. we definitely don't want private realm SSE events to leak into other sessions.